### PR TITLE
Follow up for Chaplin#780

### PR DESCRIPTION
**Pull request by Florian-R**
Not green yet. Still one test failing.

### CODE REVIEW

File: src/chaplin/views/collection_view.coffee [hunk 1]
```diff
@@ -30,7 +30,11 @@ insertView = (list, viewEl, position, length, itemSelector) ->
 
   if insertInMiddle or itemSelector
     # Get the children which originate from item views.
-    children = list.querySelectorAll(itemSelector)
+    if itemSelector
```
Comment (akre54) [inline src/chaplin/views/collection_view.coffee]: I think the Chaplin style is 

``` coffee
children = if itemSelector
    list.querySelectorAll(itemSelector)
  else
    list.children
```

**Comment by Florian-R** [comment]
Fixed the coding style but i'm starting to think it wasn't worth push this. I've seen @paulmillr suggestion to throw early just after, but after some investigations, i 'm afraid it might be a breaking change. This [jsbin](http://jsbin.com/makagaqene/1/edit?html,js,output) demonstrate visually the behavior on master of a CollectionView with an itemSelector with or without listSelector.

I've double checked with the tests on master with exos setup and this test pass, so it's definitively something supported.

The failure come from `qsa` which retrieve all the matching nodes, as opposed to the implementation on master which rely on `Backbone.utils.matchesSelector` (exposed by Exoskeleton) and filter only the directs children.

Maybe we can come up with something like, based on my previous attempt and code from Exos (quickly done, untested and ugly)

``` coffeescript
matchesSelector: (node, selector) ->
  matches = node.matches || node.webkitMatchesSelector || node.mozMatchesSelector || node.msMatchesSelector

  if !matchesSelector
    throw new Error('Element#matches is not supported');

  matches.call(node, selector)

filterChildren = (nodeList, selector) ->
  return nodeList unless selector
  for node in nodeList when matchesSelector node, selector
    if $ and $(node).is(selector)
      node
    else if matchesSelector.call(node, selector)
      node
```

**Comment by akre54** [comment]
Why not prepend the [`>` selector](https://developer.mozilla.org/en-US/docs/Web/CSS/Child_selectors) to the qsa to select only child nodes?

`list.querySelectorAll("> #{itemSelector}"`)

**Comment by Florian-R** [comment]
It could have been a nice solution indeed... But it throws an error in every browsers.

A quick googling seems to indicate there's only very contrived solutions to [use querySelectorAll with a child selector](http://stackoverflow.com/questions/3680876/use-queryselectorall-to-retrieve-direct-children/21126966).

**Comment by akre54** [comment]
Ah crap. I remember that. (It's a specific reason why we didn't move off jQuery a few years back, for instance). 

Any reason not to use [this shim](https://github.com/akre54/Backbone.NativeView/blob/0.3.2/backbone.nativeview.js#L34-L48) from Backbone.NativeView?

**Comment by akre54** [comment]
The matches on children vs listSelector isn't ideal, but I can't imagine it would break many if any things.

I'd also be fine dropping IE8 support, since almost every other browser [supports matches natively](http://caniuse.com/#feat=matchesselector).

**Comment by Florian-R** [comment]
> I'd also be fine dropping IE8 support, since almost every other browser supports matches natively.

I've mixed feeling about that. It will require a major version bump, no?

I'd rather go with the same shim than Backbone.NativeView, minus the prefixed version (excepted `msMatchesSelector` maybe).

If we go down this route, the remaining question is do we expose `matchesSelector` in Chaplin.utils (can be useful in a no jQuery world, but i'm unsure it belongs here) or do we keep it private?

**Comment by akre54** [comment]
> I've mixed feeling about that. It will require a major version bump, no?

I'd imagine any changes we make here would require a major version bump.

> I'd rather go with the same shim than Backbone.NativeView, minus the prefixed version (excepted msMatchesSelector maybe).

I dunno.... Firefox just dropped the `moz` prefix 2 versions ago, and webkit (safari and chrome both) weren't that much longer than that. It doesn't hurt anything, and we can always drop them down the line.

I'm fine exposing on `Chaplin.utils`. Would be easier to override if someone has a better shim (or our implementation doesn't work with their selector on older browsers).

**Comment by akre54** [comment]
Feel free to make that change and I'll merge

**Comment by Florian-R** [comment]
Change done, might need small tweaks to fit the coding style

> Feel free to make that change and I'll merge

My first idea was to ask you to not merge as is, as i intend to fix the tests with the Exos setup.
But as you've opened #3, maybe it's best to merge this, implement the suggested changes an then go back to fix the tests?

Sidenote. If this get merged, maybe it's a good idea to squash 1c63e3e with the last commit to keep the original PR clean.

File: src/chaplin/lib/utils.coffee [hunk 2]
```diff
@@ -2,11 +2,34 @@
...
+        #Use querySelectorAll to find all elements matching the selector,
+        #then check if the given element is included in that list.
+        #Executing the query on the parentNode reduces the resulting nodeList,
+        #(document doesn't have a parentNode).
+        nodeList = (this.parentNode || document).querySelectorAll(selector) || []
+        !!~indexOf(nodeList, this)
+
+      (node, selector) ->
+        matches.call(node, selector)
+
```
Comment (Florian-R) [inline src/chaplin/lib/utils.coffee]: Implementation is strictly the same than NativeView's one, except i've just slighly tweaked it to match the construct of  `beget`, `indexOf` etc. Let me know if you prefer to keep it like the original.

Comment (akre54) [inline src/chaplin/lib/utils.coffee]: In general I'm not a fan of functions wrapping a `call` just to preserve a signature. I think it's fine if in filterChildren we do `utils.matchesSelector.call node, selector` instead

File: src/chaplin/lib/utils.coffee [hunk 3]
```diff
@@ -2,11 +2,34 @@
...
+      ElementProto.mozMatchesSelector ||
+      ElementProto.msMatchesSelector ||
+      ElementProto.oMatchesSelector ||
+      #Make our own `Element#matches` for IE8
+      (selector) ->
+        #Use querySelectorAll to find all elements matching the selector,
+        #then check if the given element is included in that list.
+        #Executing the query on the parentNode reduces the resulting nodeList,
+        #(document doesn't have a parentNode).
+        nodeList = (this.parentNode || document).querySelectorAll(selector) || []
```
Comment (akre54) [inline src/chaplin/lib/utils.coffee]: use `@parentNode` for consistency

**Comment by Florian-R** [comment]
Should be OK now. Commited separately 969a9e3 so it can be easily reverted / squashed in the original PR.

**Pull request merged**